File: DiningPhilosophers.py
import sys, random, signal
from Philosopher import Philosopher
from Chopstick import Chopstick

colors = {
  'red': '\033[0;31m',  # red
  'yellow': '\033[33m',  # yellow
  'green': '\033[32m',  # green
  'blue': '\033[34m',  # blue
  'purple': '\033[35m'  # purple
}

c1 = Chopstick('c1')
c2 = Chopstick('c2')
c3 = Chopstick('c3')
c4 = Chopstick('c4')
c5 = Chopstick('c5')

chopsticks = [c1, c2, c3, c4, c5]

names = ['Aristotle','Confucius','Descartes','Kant','Plato']
random.shuffle(names)

p1 = Philosopher(names[0], c1, c2, colors['red'])
p2 = Philosopher(names[1], c2, c3, colors['yellow'])
p3 = Philosopher(names[2], c3, c4, colors['green'])
p4 = Philosopher(names[3], c4, c5, colors['blue'])
p5 = Philosopher(names[4], c5, c1, colors['purple'])

philosophers = [p1, p2, p3, p4, p5]

def init():
  print('Philosophers:')
  for p in philosophers:
    print(f'{p.name} ', end='')
  print('\n\nDinner is served!\n')

def start():
  for p in philosophers:
    p.start()
  for p in philosophers:
    p.join()

def exit(signum, frame):
  for p in philosophers:
    p.stop()
  print('\nDinner has ended. Waiting for everyone to finish up!\n')

if __name__ == '__main__':
  signal.signal(signal.SIGINT, exit)
  init()
  start()
  print("\nDone! 👌")

File: golang/DiningPhilosophers.go
package main

import (
	"fmt"
	"os"
	"os/signal"
	"sync"
	"syscall"
)

func main() {

	exit, wg := setup()
	<-exit
	wg.Wait()

	fmt.Println("\nSupper is over!!")
}

func setup() (chan bool, *sync.WaitGroup) {

	var wg sync.WaitGroup

	var forks []*Fork
	for i := 0; i < 5; i++ {
		forks = append(forks, newFork())
	}

	var names = [...]string{"Aristotle", "Socrates", "Descartes", "Kant", "Plato"}
	var philosophers [5]*Philosopher

	var colors = [...]string{"\033[31m",
		"\033[32m",
		"\033[33m",
		"\033[34m",
		"\033[35m"}
	var white = "\033[00m"
	for i, name := range names {
		philosophers[i] = newPhilosopher(colors[i]+name+white, forks[i], forks[(i+1)%len(names)])
	}

	for _, p := range philosophers {
		wg.Add(1)
		go p.start(&wg)
	}
	sigs := make(chan os.Signal, 1)
	exit := make(chan bool, 1)
	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)

	go func(philosophers [5]*Philosopher) {
		<-sigs
		for _, p := range philosophers {
			p.stop()
		}

		fmt.Println("\nDinner is over. Letting philosophers finish up...")
		exit <- true
	}(philosophers)

	return exit, &wg

}
